source/http2_stream_manager.c

```diff
@@ -621,6 +626,9 @@ static void s_sm_connection_on_scheduled_stream_finishes(
             aws_random_access_set_remove(&stream_manager->synced_data.ideal_available_set, sm_connection);
             work.sm_connection_to_release = sm_connection;
             --stream_manager->synced_data.holding_connections_count;
+            /* Released one connection, decrease the available concurrency by the number of streams that connection can
+             * handle */
+            stream_manager->synced_data.available_concurrency -= sm_connection->max_concurrent_streams;
```

Comment:
radical idea: what if we just grabbed a lock in the gather_metrics() function and calculated everything there

that would avoid any bugs from careful adding/subtracting at exactly the right moments

and simplify the code overall, less variables, etc